### ISSUE TITLE
Define scopes

### ISSUE THREAD

**Issue by jhass** [diaspora/api-documentation#3]
The scopes at https://github.com/diaspora/diaspora/blob/develop/app/models/api/openid_connect/authorization.rb#L21 are only preliminary, the API design should review and possibly extend and/or change these.

A couple of these are defined by the OpenID Connect specification, see http://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims The `openid` scope serves as an identifier to to disambiguate a OpenID and OAuth 2.0 request, so we have to keep it as valid. Likewise `sub` and `aud` serve as indicators in the protocol (though I didn't check if they actually need to be valid for `scope`, a quick scan through the spec doesn't make it too obvious).

There's some existing, non-concluding discussion at https://github.com/diaspora/diaspora/issues/6289

**Comment by cmrd-senya** [comment]
At diaspora/diaspora#6289 it looks like there was a consensus about having scopes for categories. Here is what I can propose:

### Read contacts scope
```
GET /api/v1/aspects
GET /api/v1/aspects/:aspect_id
GET /api/v1/aspects/:aspect_id/contacts
GET /api/v1/users/:person_guid -- for private information (for public information doesn't need a scope)
GET /api/v1/users/:person_guid/contacts
GET /api/v1/search/users -- to include hidden profiles (for publicly searchable profiles doesn't need a scope)
```

### Modify contacts scope
```
POST /api/v1/aspects
PATCH /api/v1/aspects/:aspect_id
DELETE /api/v1/aspects/:aspect_id
POST /api/v1/aspects/:aspect_id/contacts/:person_guid
DELETE /api/v1/aspects/:aspect_id/contacts/:person_guid
```

### Read private posts and interactions scope
```
GET /api/v1/posts/:post_guid
GET /api/v1/posts/:post_guid/likes
GET /api/v1/posts/:post_guid/comments
GET /api/v1/photos
GET /api/v1/photos/:photo_guid
GET /api/v1/users/:person_guid/posts
GET /api/v1/users/:person_guid/photos
GET /api/v1/search/posts -- for including private posts to the search
```

**Note**: I propose that accessing public posts should be possible without requesting any scope at all

### Publishing posts scope

```
POST /api/v1/posts
POST /api/v1/posts/:post_guid/reshares
POST /api/v1/photos
```

Making private posts would probably require access to aspects, therefore "read contacts" scope too.

### Modifying posts scope

```
DELETE /api/v1/photos/:photo_guid
DELETE /api/v1/posts/:post_guid
```

**In future will include editing as well**

### Interactions scope

```
POST /api/v1/posts/:post_guid/comments
POST /api/v1/posts/:post_guid/comments/:comment_guid/report
DELETE /api/v1/posts/:post_guid/comments/:comment_guid
POST /api/v1/posts/:post_guid/likes
DELETE /api/v1/posts/:post_guid/likes
POST /api/v1/posts/:post_guid/report
POST /api/v1/posts/:post_guid/subscribe
POST /api/v1/posts/:post_guid/mute
POST /api/v1/posts/:post_guid/hide
```

### Conversations (private messaging) scope

All conversations/messages routes.

### Read own profile scope

```
GET /api/v1/user
```

### Update own profile scope

```
PATCH /api/v1/user
```

### Read and modify tag followings scope

(everything from https://diaspora.github.io/api-documentation/routes/tag_followings.html)

```
GET /api/v1/tag_followings
POST /api/v1/tag_followings/:tag_name
DELETE /api/v1/tag_followings/:tag_name
```

### Default scope

No special permissions, available to any authenticated client without additional scopes requested.

```
GET /api/v1/users/:person_guid -- public info only
GET /api/v1/search/users -- publicly searchable profiles only
GET /api/v1/search/posts -- public posts only
GET /api/v1/posts/:post_guid/reshares

Everything from "Read private posts and interactions scope" but for public posts and interactions
```

### Streams scope

All streams scope. I would like to know if it's possible to require more than one scope for a single route. If it is possible, then I would set a requirement for two scopes for some of the stream routes.

Aspects stream = streams scope + read contacts scope
Activity stream, like stream, comment stream = streams scope + interactions scope
Tags stream = streams scope + tags scope

**Comment by cmrd-senya** [comment]
Since no objections were stated I'll start working on a pull request on the scopes

**Comment by cmrd-senya** [comment]
> The openid scope serves as an identifier to to disambiguate a OpenID and OAuth 2.0 request, so we have to keep it as valid. Likewise sub and aud serve as indicators in the protocol (though I didn't check if they actually need to be valid for scope, a quick scan through the spec doesn't make it too obvious).

`sub` and `aud` aren't scopes, but claims which are always accessed in openid authorization. Claims are user properties. In OID Connect standard scopes define access to claims.

There seems to be some confusion with our implementation because we have some claims names among scopes. That should be fixed.

So we must make `openid` scope mandatory to comply with openid connect. `openid` scope means "allow openid connect authorization".

OpenID Connect also defines 4 optional scopes: `profile email phone address`.

The optional scopes define specific user claims which are also defined by the specification.

Here are the defined claims for the `profile` scope:
` name, family_name, given_name, middle_name, nickname, preferred_username, profile, picture, website, gender, birthdate, zoneinfo, locale, and updated_at`

It is allowed to return only some of the claims. Assuming the diaspora* profiles structure the following profile claims make sense for us: `name nickname profile (= profile URL) picture gender birthdate`. We also store data for `locale` and `updated_at` which are not a part of what we call "profile" in diaspora, but I think it won't be a problem to include these data to the response too. `locale` might be useful to automatically set UI language to what user set in diaspora web settings and `updated_at` could also be displayed somewhere in client side for user's information.

We have our own API route for reading users profile and I think we can reuse the same default `profile` scope for that action too.

In OpenID Connect there is a specific scope for `address` which we could match to our location, but I think implementing this would be redundant since user can read it's own location using our API read own profile endpoint which already has it's own scope. So I think we shouldn't implement `address` scope.

We don't store phones so `phone` scope is not needed in diaspora.

A curious question is `email` scope. I think we should allow it. Email address is a part of users settings and we don't have an API route for that now so with the present API spec it won't be possible to get users email address. At the same time I think there are some useful usecases for accessing email addresses, e.g. when you want to create an account on external website using diaspora API for that (this how GitHub API is often used).

To sum up, besides the custom scopes we'll have to define I think we should support `profile` and `email` scope from OpenID Connect. For `profile` scope we support `name nickname profile (= profile URL) picture gender birthdate locale updated_at` claims and also we use the `profile` scope for our own reading profile endpoint.

`profile` scope means "Allow reading user's own profile" and `email` scope is "Allow accessing user's email address".
